fdbclient/BackupContainerAzureBlobStore.actor.cpp

```diff
@@ -0,0 +1,279 @@
+				std::istringstream iss(std::move(buffer));
+				auto resp = client->append_block_from_stream(containerName, blobName, iss).get();
+				return Void();
+			});
+		}
+		Future<int64_t> size() const override {
+			return asyncTaskThread.execAsync(
+			    [client = this->client, containerName = this->containerName, blobName = this->blobName] {
+				    auto resp = client->get_blob_properties(containerName, blobName).get().response();
+				    ASSERT(resp.valid()); // TODO: Should instead throw here
+				    return static_cast<int64_t>(resp.size);
+			    });
+		}
+		std::string getFilename() const override { return blobName; }
+		int64_t debugFD() const override { return -1; }
+	};
+
+	class BackupFile final : public IBackupFile, ReferenceCounted<BackupFile> {
+		Reference<IAsyncFile> m_file;
+
+	public:
+		BackupFile(const std::string& fileName, Reference<IAsyncFile> file) : IBackupFile(fileName), m_file(file) {}
+		Future<Void> append(const void* data, int len) override {
+			Future<Void> r = m_file->write(data, len, m_offset);
+			m_offset += len;
+			return r;
+		}
+		Future<Void> finish() override {
+			Reference<BackupFile> self = Reference<BackupFile>::addRef(this);
+			return map(m_file->sync(), [=](Void _) {
+				self->m_file.clear();
+				return Void();
+			});
+		}
+		void addref() override { ReferenceCounted<BackupFile>::addref(); }
+		void delref() override { ReferenceCounted<BackupFile>::delref(); }
+	};
+
+	static bool isDirectory(const std::string& blobName) { return blobName.size() && blobName.back() == '/'; }
+
+	ACTOR static Future<Reference<IAsyncFile>> readFile(BackupContainerAzureBlobStore* self, std::string fileName) {
+		bool exists = wait(self->blobExists(fileName));
+		if (!exists) {
+			throw file_not_found();
+		}
+		return Reference<IAsyncFile>(
+		    new ReadFile(self->asyncTaskThread, self->containerName, fileName, self->client.get()));
+	}
+
+	ACTOR static Future<Reference<IBackupFile>> writeFile(BackupContainerAzureBlobStore* self, std::string fileName) {
+		wait(self->asyncTaskThread.execAsync(
+		    [client = self->client.get(), containerName = self->containerName, fileName = fileName] {
+			    auto outcome = client->create_append_blob(containerName, fileName).get();
+			    return Void();
+		    }));
+		return Reference<IBackupFile>(
+		    new BackupFile(fileName, Reference<IAsyncFile>(new WriteFile(self->asyncTaskThread, self->containerName,
+		                                                                 fileName, self->client.get()))));
+	}
+
+	static void listFiles(AzureClient* client, const std::string& containerName, const std::string& path,
+	                      std::function<bool(std::string const&)> folderPathFilter,
+	                      BackupContainerFileSystem::FilesAndSizesT& result) {
+		auto resp = client->list_blobs_segmented(containerName, "/", "", path).get().response();
+		for (const auto& blob : resp.blobs) {
```

Comment:
If I read this right, listing files would be done serially in that there would be at most one list operation at a time outstanding to the service.  

Maybe this is fine for Azure, but FWIW the S3 version of this does a parallel recursion because it results in a much higher file listing throughput.  The folder scheme of BackupContainerFileSystem places on the order of 5,000 files in each "leaf" folder to avoid several issues when working with high file counts in folders on filesystems in general (local mounts, network mounts, S3 and clones).  Since FDB backups on large clusters and with many snapshots spread out in time can contain several million files, discovering all of the backup files for a restore's required version range could involve ~1000 folder listings or more.

With multiple background threads to handle Azure requests, it looks like you could easily parallelize the Azure BackupContainer implementation of `listFiles()` by making the function here non-recursive and moving the if/else below out to `BackupContainerAzureBlobStore::listFiles()` which would become an actor which starts from the outermost folder and lists subdirectories recursively in parallel.

Reply:
Thanks for the feedback, we will likely have to do this for Azure backups as well in a separate PR.